src/modules/credential/cert/mod.rs

```diff
@@ -0,0 +1,907 @@
+//!
+//! CA certificates are associated with a role; role names and CRL names are normalized
+//! to lower-case.
+//!
+//! Please note that to use this auth method, `tls_disable` and `tls_disable_client_certs` 
+//! must be false in the RustyVault configuration. This is because the certificates are
+//! sent through TLS communication itself.
+
+use std::sync::{Arc, RwLock};
+
+use as_any::Downcast;
+use dashmap::DashMap;
+use derive_more::Deref;
+
+use crate::{
+    core::Core,
+    errors::RvError,
+    logical::{Backend, LogicalBackend, Request, Response},
+    modules::{auth::AuthModule, Module},
+    new_logical_backend, new_logical_backend_internal,
+};
+
+pub mod path_certs;
+pub mod path_config;
+pub mod path_crls;
+pub mod path_login;
+
+pub use path_certs::CertEntry;
+pub use path_crls::CRLInfo;
+
+static CERT_BACKEND_HELP: &str = r#"
+The "cert" credential provider allows authentication using
+TLS client certificates. A client connects to RustyVault and uses
+the "login" endpoint to generate a client token.
+
+Trusted certificates are configured using the "certs/" endpoint
+by a user with root access. A certificate authority can be trusted,
+which permits all keys signed by it. Alternatively, self-signed
+certificates can be trusted avoiding the need for a CA.
+"#;
+
+pub struct CertModule {
+    pub name: String,
+    pub backend: Arc<CertBackend>,
+}
+
+pub struct CertBackendInner {
+    pub core: Arc<RwLock<Core>>,
+    pub crls: DashMap<String, CRLInfo>,
+}
+
+#[derive(Deref)]
+pub struct CertBackend {
+    #[deref]
+    pub inner: Arc<CertBackendInner>,
+}
+
+impl CertBackend {
+    pub fn new(core: Arc<RwLock<Core>>) -> Self {
+        let inner = CertBackendInner { core, crls: DashMap::new() };
+        Self { inner: Arc::new(inner) }
+    }
+
+    pub fn new_backend(&self) -> LogicalBackend {
+        let cert_backend_ref = Arc::clone(&self.inner);
+
+        let mut backend = new_logical_backend!({
+            unauth_paths: ["login"],
+            auth_renew_handler: cert_backend_ref.renew_path_login,
+            help: CERT_BACKEND_HELP,
+        });
+
+        backend.paths.push(Arc::new(self.config_path()));
+        backend.paths.push(Arc::new(self.certs_path()));
+        backend.paths.push(Arc::new(self.certs_list_path()));
+        backend.paths.push(Arc::new(self.crl_path()));
+        backend.paths.push(Arc::new(self.crl_list_path()));
+        backend.paths.push(Arc::new(self.login_path()));
+
+        backend
+    }
+}
+
+impl CertBackendInner {
+    pub fn renew_path_login(&self, _backend: &dyn Backend, _req: &mut Request) -> Result<Option<Response>, RvError> {
+        Ok(None)
+    }
+}
+
+impl CertModule {
+    pub fn new(core: &Core) -> Self {
+        Self {
+            name: "cert".to_string(),
+            backend: Arc::new(CertBackend::new(Arc::clone(core.self_ref.as_ref().unwrap()))),
+        }
+    }
+}
+
+impl Module for CertModule {
+    fn name(&self) -> String {
+        return self.name.clone();
```

Comment:
> It seems CertModule.name is always a constant string through the whole lifetime of a RustyVault instance, then why don't return a referent &string instead of cloning a new string?

This field and function was originally used for debugging or logging. Using references with lifecycles is relatively complex. For simplicity, String was used. Later, let's see whether this field and function should be removed or optimized.